### ISSUE TITLE
Inference tests do not catch problems with thunked inputs

### ISSUE THREAD

**Issue by mzgubic** [JuliaDiff/ChainRulesTestUtils.jl#214]
Consider:

```julia
julia> my_id(x) = x
my_id (generic function with 1 method)

julia> function ChainRulesCore.rrule(::typeof(my_id), x)
           my_id_pb(ȳ) = (NoTangent(), ȳ)
           function my_id_pb(ȳ::AbstractThunk)
               precision = rand() > 0.5 ? Float64 : Float32
               return (NoTangent(), precision(unthunk(ȳ)))
           end
           return x, my_id_pb
       end

julia> test_rrule(my_id, 2.0)
Test Summary:                | Pass  Total
test_rrule: my_id on Float64 |    7      7
Test.DefaultTestSet("test_rrule: my_id on Float64", Any[], 7, false, false)
```

while really we should be catching that the thunked cotangents result in type unstable pullbacks. This may seem obscure but is quite relevant because of [closures being unstable if they call another method of themselves](https://juliadiff.org/ChainRulesCore.jl/dev/writing_good_rules.html#Be-careful-about-pullback-closures-calling-other-methods-of-themselves).
